static/js/stream_list.js

```diff
@@ -121,63 +103,40 @@ exports.build_stream_list = function () {
         return;
     }
 
-    streams = filter_streams_by_search(streams);
+    // The main logic to build the list is in stream_sort.js, and
+    // we get thre lists of streams (pinned/normal/dormant).
+    var stream_groups = stream_sort.sort_groups(get_search_term());
+
+    if (stream_groups.same_as_before) {
+        return;
+    }
 
-    var sort_recent = (streams.length > 40);
-    var pinned_streams = [];
-    var unpinned_streams = [];
     var parent = $('#stream_filters');
     var elems = [];
 
     function add_sidebar_li(stream) {
         var sub = stream_data.get_sub(stream);
         var sidebar_row = exports.stream_sidebar.get_row(sub.stream_id);
-        if (sort_recent) {
-            sidebar_row.update_whether_active();
-        }
+        sidebar_row.update_whether_active();
         elems.push(sidebar_row.get_li().get(0));
     }
 
-    _.each(streams, function (stream) {
-        var pinned = stream_data.get_sub(stream).pin_to_top;
-        if (pinned) {
-            pinned_streams.push(stream);
-        } else {
-            unpinned_streams.push(stream);
-        }
-    });
+    parent.empty();
 
-    pinned_streams.sort(util.strcmp);
+    _.each(stream_groups.pinned_streams, add_sidebar_li);
 
-    unpinned_streams.sort(function (a, b) {
-        if (sort_recent) {
-            if (stream_data.is_active(b) && ! stream_data.is_active(a)) {
-                return 1;
-            } else if (! stream_data.is_active(b) && stream_data.is_active(a)) {
-                return -1;
-            }
-        }
-        return util.strcmp(a, b);
-    });
+    if (stream_groups.pinned_streams.length > 0) {
+        elems.push($('<hr class="stream-split">').get(0));
+    }
 
-    streams = pinned_streams.concat(unpinned_streams);
+    _.each(stream_groups.normal_streams, add_sidebar_li);
 
-    if (previous_sort_order !== undefined &&
-        util.array_compare(previous_sort_order, streams) &&
-        util.array_compare(previous_unpinned_order, unpinned_streams)) {
-        return;
+    if (stream_groups.dormant_streams.length > 0) {
+        elems.push($('<hr class="stream-split">').get(0));
     }
-    previous_sort_order = streams;
-    previous_unpinned_order = unpinned_streams;
-    parent.empty();
 
-    if (pinned_streams.length > 0) {
-        _.each(pinned_streams, add_sidebar_li);
-        elems.push($('<hr class="pinned-stream-split">').get(0));
-    }
-    if (unpinned_streams.length > 0) {
-        _.each(unpinned_streams, add_sidebar_li);
-    }
+
+    _.each(stream_groups.dormant_streams, add_sidebar_li);
```

Comment:
What happens when a pinned stream is dormant?  I guess we treat it as just pinned, which is probably right.

(BTW, might make sense to rename "inactive" to "dormant" in the CSS; I like the `dormant` name).